python/ray/data/dataset.py

```diff
@@ -808,26 +815,11 @@ def randomize_block_order(
                 based on system randomness.
 
         Returns:
-            The shuffled dataset.
+            The block-shuffled dataset.
         """
 
-        def do_randomize_block_order(block_list, *_):
-            num_blocks = block_list.executed_num_blocks()  # Blocking.
-            if num_blocks == 0:
-                return block_list, {}
-
-            randomized_block_list = block_list.randomize_block_order(seed)
-
-            return randomized_block_list, {}
-
-        plan = self._plan.with_stage(
-            AllToAllStage(
-                "randomize_block_order",
-                None,
-                do_randomize_block_order,
-            )
-        )
-        return Dataset(plan, self._epoch, self._lazy)
+        plan = self._plan.with_stage(RandomizeBlocksStage(seed))
+        return Dataset(plan, self._epoch, self._lazy, defer_execution=True)
```

Comment:
Hmm having both `lazy` and `defer_execution` is a bit odd, why can't this follow the `self._lazy` semantics? Is there an issue with executing this stage eagerly?

Reply:
Good point. I can't remember why I did this... removed it.

Reply:
Oh, it breaks lazy read->map_batches() fusion, that's why. I reverted this since it broke a unit test.

Reply:
Hmm do you know why?